python/adt.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+"""
+Algebraic Data Types (ADT) is used to represent two kinds of things:
```

Comment:
This is the first way I considered doing ADTs in python too, but every professional python programmer I knew told me that discriminating using type() or similar functions for any purpose other than optimization is considered in extremely poor form.

This may be OK anyways, but the python folks are unlikely to be happy with this representation.